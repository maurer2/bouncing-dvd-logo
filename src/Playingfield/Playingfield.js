import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';

import random from 'lodash.random';

import { StyleSheetManager } from 'styled-components/macro';
import Logo from '../Logo/Logo';
import Sound from '../Sound/Sound';
import * as Styles from './Playingfield.styles';

const isPastStartBoundary = (position) => (position <= 0);

const isPastEndBoundary = (position, objectSize, playfieldSize) => {
  const maxPositionStillInside = (playfieldSize - objectSize);

  return position >= maxPositionStillInside;
};

/*
const isCollidingWithBoundaries = (
  positionX, positionY, width, height, playfieldWidth, playfieldHeight,
) => {
  const leftCheck = isPastStartBoundary(positionX);
  const rightCheck = isPastEndBoundary(positionX, width, playfieldWidth);
  const topCheck = isPastStartBoundary(positionY);
  const bottomCheck = isPastEndBoundary(positionY, height, playfieldHeight);

  return [leftCheck, rightCheck, topCheck, bottomCheck].some(entry => !!entry);
};
*/

const PlayField = (props) => {
  const [positionX, setPositionX] = useState(0);
  const [positionY, setPositionY] = useState(0);

  const loopTimestamp = useRef(0);
  const changeDeltaX = useRef(2); // velocity x
  const changeDeltaY = useRef(2); // velocity y

  const playfieldDomElement = useRef();
  const playfieldBB = useRef();

  const isColliding = useRef(false);
  const isPaused = useRef(false);
  const isInit = useRef(false);
  const maxRandomness = 6; // max value of deviation from correct reflection on collision

  // svg
  const width = 150;
  const height = 138; // AR 0,92

  // set random initial position and direction
  function initPosition() {
    const { width: widthBB, height: heightBB } = playfieldBB.current;

    setPositionX(() => random(widthBB - width));
    setPositionY(() => random(heightBB - height));

    // initial direction
    changeDeltaX.current = random(1) === 0 ? changeDeltaX.current * -1 : changeDeltaX.current * +1;
    changeDeltaY.current = random(1) === 0 ? changeDeltaY.current * -1 : changeDeltaY.current * +1;
    isInit.current = true;
  }

  function updatePosition() {
    const { width: widthBB, height: heightBB } = playfieldBB.current;

    const upperRandomBound = 1.0 + ((maxRandomness / 2) / 100);
    const lowerRandomBound = 1.0 - ((maxRandomness / 2) / 100);

    let newChangeDeltaX = changeDeltaX.current * random(lowerRandomBound, upperRandomBound, true);
    let newChangeDeltaY = changeDeltaY.current * random(lowerRandomBound, upperRandomBound, true);
    let hasCollided = false;

    setPositionX((prevPositionX) => {
      if (isPastStartBoundary(prevPositionX)) {
        newChangeDeltaX = Math.abs(changeDeltaX.current);
        hasCollided = true;
      }

      if (isPastEndBoundary(prevPositionX, width, widthBB)) {
        newChangeDeltaX = Math.abs(changeDeltaX.current) * -1;
        hasCollided = true;
      }

      return Math.round(prevPositionX + newChangeDeltaX);
    });

    setPositionY((prevPositionY) => {
      if (isPastStartBoundary(prevPositionY)) {
        newChangeDeltaY = Math.abs(changeDeltaY.current);
        hasCollided = true;
      }

      if (isPastEndBoundary(prevPositionY, height, heightBB)) {
        newChangeDeltaY = Math.abs(changeDeltaY.current) * -1;
        hasCollided = true;
      }

      return Math.round(prevPositionY + newChangeDeltaY);
    });

    if (hasCollided) {
      changeDeltaX.current = newChangeDeltaX;
      changeDeltaY.current = newChangeDeltaY;
    }

    isColliding.current = hasCollided;
  }

  function loop() {
    if (!isPaused.current) {
      updatePosition();
    }

    loopTimestamp.current = window.requestAnimationFrame(loop);
  }

  function startLoop() {
    if (loopTimestamp.current !== 0) {
      return;
    }

    loopTimestamp.current = window.requestAnimationFrame(loop);
  }

  function stopLoop() {
    window.cancelAnimationFrame(loopTimestamp.current);
  }

  useEffect(() => {
    playfieldBB.current = playfieldDomElement.current.getBoundingClientRect();

    initPosition();
    startLoop();

    return () => stopLoop();
  }, []);

  useEffect(() => {
    isPaused.current = props.isPaused;
  }, [props.isPaused]);

  return (
    <StyleSheetManager disableVendorPrefixes>
      <Styles.PlayingFieldWrapper ref={ playfieldDomElement }>
        {isInit.current && (
          <>
            <Logo
              positionX={ positionX }
              positionY={ positionY }
              width={ width }
              height={ height }
              changeColours={ isColliding.current }
            />
            <Sound playSound={ isColliding.current } />
          </>
        )}
      </Styles.PlayingFieldWrapper>
    </StyleSheetManager>
  );
};

const { bool } = PropTypes;

PlayField.propTypes = {
  isPaused: bool,
};

export default PlayField;
