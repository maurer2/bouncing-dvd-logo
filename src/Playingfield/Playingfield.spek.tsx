import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import Store, { colours } from '../Store';

import Component from './Playingfield';
import type { PlayingfieldProps } from './Playingfield.types';

const mockRandom = 10;
vi.mock('lodash-es', () => ({ random: vi.fn().mockImplementation(() => mockRandom) }));

expect(true).toBe(true)

// describe('Playingfield', () => {
//   beforeEach(() => {
//     mockRandom = 10;
//   });

//   const storeValues = {
//     colours: [...colours],
//     soundIsDisabled: true,
//     toggleSound: vi.fn(),
//   };

//   const defaultProps: PlayingfieldProps = { isPaused: false };

//   const StoreProvider = ({ children }) => (
//     <Store.Provider value={storeValues}>{children}</Store.Provider>
//   );

//   const setup = (props) =>
//     render(
//       <StoreProvider key={props?.key}>
//         <Component
//           {...defaultProps}
//           {...props}
//         />
//       </StoreProvider>,
//     );

//   it('should render ', () => {
//     const screen = setup({});

//     expect(screen.getByTestId('logo-element')).toBeInTheDocument();
//   });

//   it('should match snapshot', () => {
//     const screen = setup({});

//     expect(screen.container.firstChild).toMatchSnapshot();
//   });

//   it('should have random start value', () => {
//     mockRandom = 25;

//     const screen = setup({});

//     expect(screen.getByTestId('logo-element').style.getPropertyValue('transform')).toBe(
//       `translate(${mockRandom}px, ${mockRandom}px)`,
//     );
//   });

//   it('should have no paused attribute when game is running', () => {
//     const props: PlayingfieldProps = { isPaused: false };
//     const screen = setup(props);

//     expect(screen.getByTestId('playfingfield')).toHaveAttribute('data-status', 'active');

//     expect(screen.container.firstChild).toMatchSnapshot();
//   });

//   it('should have paused attribute when game is paused', () => {
//     const props: PlayingfieldProps = { isPaused: true };
//     const screen = setup(props);

//     expect(screen.getByTestId('playfingfield')).toHaveAttribute('data-status', 'inactive');

//     expect(screen.container.firstChild).toMatchSnapshot();
//   });
// });
