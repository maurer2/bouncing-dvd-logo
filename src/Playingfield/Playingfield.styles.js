import styled from 'styled-components/macro';

// eslint-disable-next-line import/prefer-default-export
export const PlayingFieldWrapper = styled.div`
  position: relative;
  width: 100%;
  height: 100%;
  pointer-events: none;
  transform: translateZ(0);
`;
