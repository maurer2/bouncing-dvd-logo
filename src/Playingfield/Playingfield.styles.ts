import styled from 'styled-components';

export const PlayingFieldWrapper = styled.div`
  position: relative;
  width: 100%;
  height: 100%;
  transform: translateZ(0);
`;
