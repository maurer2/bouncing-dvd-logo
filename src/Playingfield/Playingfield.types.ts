import type { Colour } from '../Store/types';

export type PlayingfieldProps = {
  isPaused: boolean;
  triggerCollision: () => void;
  currentColor: Colour;
};

export type LogoDimensions = [width: number, height: number];

export type PositionAndVelocity = {
  value: number;
  velocity: number;
  randomness: number;
};

export type ReducerState = {
  positionX: PositionAndVelocity;
  positionY: PositionAndVelocity;
};
