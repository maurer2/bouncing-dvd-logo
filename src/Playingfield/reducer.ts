import type * as Types from './Playingfield.types';

export const reducer = (state: Types.ReducerState, action: Types.ReducerAction): void => {
  switch (action.type) {
    case 'TRIGGER_INITIAL_POSITION': {
      const newDraft = state;
      const { worldSize, logoSize, velocityX, velocityY } = action.payload;

      newDraft.positionX.value = worldSize.width / 2 - logoSize[0] / 2; // dead centre
      newDraft.positionY.value = worldSize.height / 2 - logoSize[1] / 2; // dead centre
      newDraft.positionX.velocity = velocityX;
      newDraft.positionY.velocity = velocityY;

      break;
    }
    case 'TRIGGER_NEXT_POSITION': {
      const newDraft = state;

      if (newDraft.positionX.value === null || newDraft.positionY.value === null) {
        return;
      }

      newDraft.positionX.value += state.positionX.velocity;
      newDraft.positionY.value += state.positionY.velocity;
      break;
    }
    case 'TRIGGER_X_COLLISION': {
      const newDraft = state;
      const velocityX = action.payload;

      if (newDraft.positionX.value === null) {
        return;
      }

      newDraft.positionX.value += velocityX;
      newDraft.positionX.velocity = velocityX;
      break;
    }
    case 'TRIGGER_Y_COLLISION': {
      const newDraft = state;
      const velocityY = action.payload;

      if (newDraft.positionY.value === null) {
        return;
      }

      newDraft.positionY.value += velocityY;
      newDraft.positionY.velocity = velocityY;
      break;
    }
    default: {
      break;
    }
  }
};
