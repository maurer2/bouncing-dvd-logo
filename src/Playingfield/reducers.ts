import { produce } from 'immer';

type PositionAndVelocity = {
  value: number | null;
  velocity: number;
};

export type ReducerState = {
  positionX: PositionAndVelocity;
  positionY: PositionAndVelocity;
};

export type TriggerInitialPosition = {
  type: 'TRIGGER_INITIAL_POSITION';
  payload: {
    worldSize: Pick<DOMRectReadOnly, 'width' | 'height'>;
    logoSize: [width: number, height: number];
    velocityX: number;
    velocityY: number;
  };
};

export type TriggerNextPosition = {
  type: 'TRIGGER_NEXT_POSITION';
};

export type TriggerXCollision = {
  type: 'TRIGGER_X_COLLISION';
  payload: number;
};

export type TriggerYCollision = {
  type: 'TRIGGER_Y_COLLISION';
  payload: number;
};

export type ReducerAction =
  | TriggerInitialPosition
  | TriggerNextPosition
  | TriggerXCollision
  | TriggerYCollision;

// https://prateeksurana.me/blog/simplify-immutable-data-structures-in-usereducer-with-immer/#:~:text=Thunder%27%2C%20%27Virtually%20Immortal%27%5D-,useReducer%20with%20Immer,-Now%20that%20we
export const reducers = produce((state: ReducerState, action: ReducerAction) => {
  switch (action.type) {
    case 'TRIGGER_INITIAL_POSITION': {
      const newDraft = state;
      const { worldSize, logoSize, velocityX, velocityY } = action.payload;

      newDraft.positionX.value = worldSize.width / 2 - logoSize[0] / 2; // dead centre
      newDraft.positionY.value = worldSize.height / 2 - logoSize[1] / 2; // dead centre
      newDraft.positionX.velocity = velocityX;
      newDraft.positionY.velocity = velocityY;

      break;
    }
    case 'TRIGGER_NEXT_POSITION': {
      const newDraft = state;

      if (newDraft.positionX.value === null || newDraft.positionY.value === null) {
        return;
      }

      newDraft.positionX.value = Math.round(newDraft.positionX.value + state.positionX.velocity);
      newDraft.positionY.value = Math.round(newDraft.positionY.value + state.positionY.velocity);

      break;
    }
    case 'TRIGGER_X_COLLISION': {
      const newDraft = state;
      const velocityX = action.payload;

      if (newDraft.positionX.value === null) {
        return;
      }

      newDraft.positionX.value = Math.round(newDraft.positionX.value + velocityX);
      newDraft.positionX.velocity = Math.round(velocityX);

      break;
    }
    case 'TRIGGER_Y_COLLISION': {
      const newDraft = state;
      const velocityY = action.payload;

      if (newDraft.positionY.value === null) {
        return;
      }

      newDraft.positionY.value = Math.round(newDraft.positionY.value + velocityY);
      newDraft.positionY.velocity = Math.round(velocityY);

      break;
    }
    default: {
      break;
    }
  }
});
