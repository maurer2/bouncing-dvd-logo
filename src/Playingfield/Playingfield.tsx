import React, {
  useEffect,
  useState,
  useRef,
  useCallback,
  VFC,
  useLayoutEffect,
} from 'react';
import PropTypes from 'prop-types';
import { random } from 'lodash-es';

import Logo from '../Logo/Logo';
import Sound from '../Sound/Sound';
import Controls from '../Controls/Controls';
import useChangeDelta from '../Hooks/useChangeDelta';
import useCollisionDetection from '../Hooks/useCollisionDetection';

import * as Styles from './Playingfield.styles';
import * as Types from './Playingfield.types';

const PlayingField: VFC<Readonly<Types.PlayingfieldProps>> = ({ isPaused }): JSX.Element => {
  const [positionX, setPositionX] = useState(0);
  const [positionY, setPositionY] = useState(0);

  const loopTimestamp = useRef(0);
  const playfieldBB = useRef<DOMRect>({} as DOMRect);

  const isColliding = useRef(false);
  const isCollidingX = useRef(false);
  const isCollidingY = useRef(false);
  const isPausedPrevious = useRef(false);
  const isInit = useRef(false);

  // svg
  const widthObject = 150;
  const heightObject = 138; // AR 0,92

  const playfieldDomRefCB = useCallback((element: HTMLElement) => {
    if (element === null) {
      return;
    }

    playfieldBB.current = element.getBoundingClientRect();
  }, []);

  const [isCollidingXStart, isCollidingXEnd] = useCollisionDetection(
    positionX,
    widthObject,
    playfieldBB.current.width,
  );
  const [isCollidingYStart, isCollidingYEnd] = useCollisionDetection(
    positionY,
    heightObject,
    playfieldBB.current.height,
  );
  const [changeX] = useChangeDelta(3, isCollidingX.current);
  const [changeY] = useChangeDelta(3, isCollidingY.current);

  // set random initial position and direction
  // useEffectOnce
  const initPosition = useCallback(() => {
    if (isInit.current) {
      return;
    }

    const { width, height } = playfieldBB.current;

    setPositionX(() => random(width - widthObject));
    setPositionY(() => random(height - heightObject));

    isInit.current = true;
  }, []);

  const loop = useCallback(() => {
    if (!isPausedPrevious.current) {
      let hasCollidedX = false;
      let hasCollidedY = false;

      setPositionX((prevPositionX) => {
        if (isCollidingXStart.current || isCollidingXEnd.current) {
          hasCollidedX = true;
        }

        return Math.round(prevPositionX + changeX.current);
      });

      setPositionY((prevPositionY) => {
        if (isCollidingYStart.current || isCollidingYEnd.current) {
          hasCollidedY = true;
        }

        return Math.round(prevPositionY + changeY.current);
      });

      isColliding.current = hasCollidedX || hasCollidedY;
      isCollidingX.current = hasCollidedX;
      isCollidingY.current = hasCollidedY;
    }

    loopTimestamp.current = window.requestAnimationFrame(loop);
  }, [isCollidingXStart, isCollidingXEnd, isCollidingYStart, isCollidingYEnd, changeX, changeY]);

  const startLoop = useCallback(() => {
    if (loopTimestamp.current !== 0) {
      return;
    }

    loopTimestamp.current = window.requestAnimationFrame(loop);
  }, [loop]);

  function stopLoop(): void {
    window.cancelAnimationFrame(loopTimestamp.current);
  }

  useLayoutEffect(() => {
    initPosition();
    startLoop();

    return () => stopLoop();
  }, [initPosition, startLoop]);

  useEffect(() => {
    isPausedPrevious.current = isPaused;
  }, [isPaused]);

  return (
    <Styles.PlayingFieldWrapper
      ref={playfieldDomRefCB}
      data-testid="playfingfield"
      data-status={isPaused ? 'inactive' : 'active'}
    >
      {isInit.current && (
        <>
          <Logo
            positionX={positionX}
            positionY={positionY}
            width={widthObject}
            height={heightObject}
            changeColours={isColliding.current}
            isPaused={isPaused}
          />
          <Controls />
          <Sound triggerSound={isColliding.current} />
        </>
      )}
    </Styles.PlayingFieldWrapper>
  );
};

const { bool } = PropTypes;

PlayingField.propTypes = { isPaused: bool.isRequired };

export default PlayingField;
