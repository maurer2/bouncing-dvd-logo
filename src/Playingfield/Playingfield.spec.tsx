import React from 'react';
import '@testing-library/jest-dom';
import { render, within } from '@testing-library/react';
import 'jest-styled-components';

import Store, { colours } from '../Store';

import Component from './Playingfield';
import { PlayingfieldProps } from './Playingfield.types';

let mockRandom = 10;
jest.mock('lodash-es', () => ({ random: jest.fn().mockImplementation(() => mockRandom) }));

describe('Playingfield', () => {
  beforeEach(() => {
    // mockRandom = jest.spyOn(global.Math, 'random');
    // mockRandom.mockReturnValue(0.123456);
    mockRandom = 10;
  });

  const storeValues = {
    colours: [...colours],
    soundIsDisabled: true,
    toggleSound: jest.fn(),
  };

  const defaultProps: PlayingfieldProps = { isPaused: false };

  const StoreProvider = ({ children }) => (
    <Store.Provider value={storeValues}>{children}</Store.Provider>
  );

  const setup = (props) => render(
    <StoreProvider key={props?.key}>
      <Component {...defaultProps} {...props} />
    </StoreProvider>,
  );

  it('should render ', () => {
    const screen = setup({});

    expect(screen.getByTestId('logo-element')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const screen = setup({});

    expect(screen.container.firstChild).toMatchSnapshot();
  });

  it('should have random start value', () => {
    mockRandom = 25;

    const screen = setup({});

    expect(screen.getByTestId('logo-element').style.getPropertyValue('transform')).toBe(
      `translate(${mockRandom}px, ${mockRandom}px)`,
    );
  });

  it('should have no paused attribute when game is running', () => {
    const props: PlayingfieldProps = { isPaused: false };
    const screen = setup(props);

    expect(screen.getByTestId('playfingfield')).toHaveAttribute('data-status', 'active');

    expect(screen.container.firstChild).toMatchSnapshot();
  });

  it('should have paused attribute when game is paused', () => {
    const props: PlayingfieldProps = { isPaused: true };
    const screen = setup(props);

    expect(screen.getByTestId('playfingfield')).toHaveAttribute('data-status', 'inactive');

    expect(screen.container.firstChild).toMatchSnapshot();
  });
});
