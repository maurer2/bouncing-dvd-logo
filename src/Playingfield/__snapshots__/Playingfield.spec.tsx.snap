// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playingfield should have no paused attribute when game is running 1`] = `
.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 150px;
  height: 138px;
  margin: 0;
  will-change: transform;
  opacity: 1;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  bottom: 0.5rem;
  right: 0.5rem;
}

.c3 {
  padding: 0;
  border: 0;
  background: none;
  color: white;
  outline: none;
  cursor: pointer;
  width: 5vw;
  height: 5vw;
  min-width: 50px;
  min-height: 50px;
}

.c4 {
  display: block;
  fill: currentColor;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c4 .sound-waves {
  display: none;
}

.c0 {
  position: relative;
  width: 100%;
  height: 100%;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

<div
  class="c0"
  data-status="active"
  data-testid="playfingfield"
>
  <figure
    class="c1"
    data-testid="logo-element"
    style="transform: translate(10px, 10px); color: white;"
  >
    <svg
      data-testid="cat-logo"
    >
      SVG markup
    </svg>
  </figure>
  <div
    class="c2"
    data-testid="controls"
  >
    <button
      class="c3"
      data-testid="controls-control"
      type="button"
    >
      <svg
        class="c4"
        data-testid="controls-icon"
        status="inactive"
      >
        SVG markup
      </svg>
    </button>
  </div>
</div>
`;

exports[`Playingfield should have paused attribute when game is paused 1`] = `
.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 150px;
  height: 138px;
  margin: 0;
  will-change: transform;
  opacity: 0.25;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  bottom: 0.5rem;
  right: 0.5rem;
}

.c3 {
  padding: 0;
  border: 0;
  background: none;
  color: white;
  outline: none;
  cursor: pointer;
  width: 5vw;
  height: 5vw;
  min-width: 50px;
  min-height: 50px;
}

.c4 {
  display: block;
  fill: currentColor;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c4 .sound-waves {
  display: none;
}

.c0 {
  position: relative;
  width: 100%;
  height: 100%;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

<div
  class="c0"
  data-status="inactive"
  data-testid="playfingfield"
>
  <figure
    class="c1"
    data-testid="logo-element"
    style="transform: translate(10px, 10px); color: white;"
  >
    <svg
      data-testid="cat-logo"
    >
      SVG markup
    </svg>
  </figure>
  <div
    class="c2"
    data-testid="controls"
  >
    <button
      class="c3"
      data-testid="controls-control"
      type="button"
    >
      <svg
        class="c4"
        data-testid="controls-icon"
        status="inactive"
      >
        SVG markup
      </svg>
    </button>
  </div>
</div>
`;

exports[`Playingfield should match snapshot 1`] = `
.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 150px;
  height: 138px;
  margin: 0;
  will-change: transform;
  opacity: 1;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  bottom: 0.5rem;
  right: 0.5rem;
}

.c3 {
  padding: 0;
  border: 0;
  background: none;
  color: white;
  outline: none;
  cursor: pointer;
  width: 5vw;
  height: 5vw;
  min-width: 50px;
  min-height: 50px;
}

.c4 {
  display: block;
  fill: currentColor;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c4 .sound-waves {
  display: none;
}

.c0 {
  position: relative;
  width: 100%;
  height: 100%;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

<div
  class="c0"
  data-status="active"
  data-testid="playfingfield"
>
  <figure
    class="c1"
    data-testid="logo-element"
    style="transform: translate(10px, 10px); color: white;"
  >
    <svg
      data-testid="cat-logo"
    >
      SVG markup
    </svg>
  </figure>
  <div
    class="c2"
    data-testid="controls"
  >
    <button
      class="c3"
      data-testid="controls-control"
      type="button"
    >
      <svg
        class="c4"
        data-testid="controls-icon"
        status="inactive"
      >
        SVG markup
      </svg>
    </button>
  </div>
</div>
`;
